### DISCUSSION TITLE
Project Setup Libs

### DISCUSSION THREAD

**Discussion by qnotna** [TRNO-ENTERPRISE/treno-game#7] [Foundations]
Welchen Setup wollen wir nutzen, damit wir organisiert unabhängig voneinander arbeiten können?

[cpplint](https://github.com/cpplint/cpplint) scheint weit verbreitet, dadurch können wir sicherstellen dass der Code immer gleich formattiert ist, egal wer Commits macht. Google bspw. hat einen eigenen [Styleguide für cpplint](https://google.github.io/styleguide/cppguide.html) dem interne C++ Projekte folgen

CMake werden wir ja bestimmt auch nutzen müssen

Einige Package Manager die ich auf SO gefunden hab: 
* [bpt](https://github.com/vector-of-bool/dds)
* [cpm](http://www.cpm.rocks/)
* [conan](https://conan.io/)
* [poac](https://github.com/poacpm/poac)
* [pacm](http://sourcey.com/pacm/)
* [spack](https://spack.io/)
* [buckaroo](http://buckaroo.pm/)
* [hunter](https://github.com/ruslo/hunter)
* [vcpkg](https://github.com/Microsoft/vcpkg)

**Comment by Postea** [comment]
# CMake

## Allgemein
CMake ist ein Code Management Tool. Es sehr gut kompatibel mit VSCode und die Projektstruktur von VSCode. Außerdem ist es mit der Make-File-Syntax breit anwendbar, also werden unsere Erkenntnisse, die wir mit CMake gewinnen auch gut anderswo anwenden können.

## Relevanz
CMake ist relevant für unser Projekt, da es uns eine einheitliche Organisation gibt. Es gibt uns, sowohl die Möglichkeit unseren Code strukturell zu kompilieren, als auch verschiedene Umgebungen zu generieren, zb. TEST, PROD, als auch erlaubt es uns externe Bibliotheken in unser Projekt einzubinden. Wir können für ein entspannteres Handling von Packages auch [CPM ](https://github.com/cpm-cmake/CPM.cmake) verwenden. Dies ist ein sehr lightweight layer auf CMake, der noch zusätzlich Dependency Management übernimmt.


## Beispiele

### CMake
Beispiel eines CMake-Files, wo wir ein Main-File in das Projekt einbinden:
```
cmake_minimum_required(VERSION 3.10)
project(Tutorial)
add_executable(Tutorial tutorial.cxx)
```
Nun können wir in dem Verzeichnis unseres Projektes den Befehl `cmake --build .` ausführen und unser Projekt ist kompiliert

### CPM
```
cmake_minimum_required(VERSION 3.14 FATAL_ERROR)

# create project
project(MyProject)

# add executable
add_executable(main main.cpp)

# add dependencies
include(cmake/CPM.cmake)

CPMAddPackage("gh:fmtlib/fmt#7.1.3")
CPMAddPackage("gh:nlohmann/json@3.10.5")
CPMAddPackage("gh:catchorg/Catch2@2.5.0")

# link dependencies
target_link_libraries(main fmt::fmt nlohmann_json::nlohmann_json Catch2::Catch2)
```

Nachdem CPM zum Projekt hinzugefügt wurde ( `include(cmake/CPM.cmake)` ) kann man nun die CPM-Befehle im Makefile verwenden.

## Links
https://cmake.org/cmake/help/latest/guide/tutorial/A%20Basic%20Starting%20Point.html#exercise-1-building-a-basic-project
https://github.com/cpm-cmake/CPM.cmake

**Comment by cancoel** [comment]
# vcpkg (Package Manager)

## Allgemein
- Open source C/C++ package manager von Microsoft
- Einfache Nutzung und plattformunabhängig
- Integration in Visual Studio, Visual Studio Code und andere IDEs (JetBrains...) möglich
- Funktioniert mit CMake, siehe https://vcpkg.io/en/docs/users/buildsystems/cmake-integration.html

## Relevanz
- Wird laufend weiterentwickelt
- Bietet Workflows für Gruppen-Projekte

## Beispiele
...

## Links
https://vcpkg.io/en/index.html
